### PR TITLE
Small Pinning API improvements

### PR DESCRIPTION
**Pull request by dave-tucker**
1, Allow pinning from the program directly (e.g `TracePoint::pin("/path")`
2. Add `FdLink::pin("/path")`, which requires that you've called `take_link()` first, which actually isn't a bad thing at all :wink:

### CODE REVIEW

File: aya/src/programs/links.rs [hunk 1]
```diff
@@ -111,6 +137,24 @@ impl FdLink {
...
         FdLink { fd }
     }
+
+    /// Pins the FdLink to a BPF filesystem.
+    ///
+    /// When a BPF object is pinned to a BPF filesystem it will remain attached after
+    /// Aya has detached the link.
+    /// To remove the attachment, the file on the BPF filesystem must be remove.
+    /// Any directories in the the path provided should have been created by the caller.
+    pub fn pin<P: AsRef<Path>>(&self, path: P) -> Result<(), LinkError> {
```
Comment (alessandrod) [inline aya/src/programs/links.rs:147]: Why is this implemented on `FdLink`? Can other links not be pinned?

**Comment by dave-tucker** [comment]
Comments addressed. All pin operations across Map, Program and Links now use `PinError`.
Link pinning is restricted to FdLink because, as far as I know, only BPF FDs can be pinned, which is not what is returned by other link types.

See: #329 which uses `bpf_link_create` for programs that used PerfLink, which would now allow for tracepoints, kprobes etc.. to be pinned.

File: aya/src/maps/mod.rs [hunk 2]
```diff
@@ -315,10 +301,12 @@ impl Map {
                 })
             }
         };
```
Comment (alessandrod) [inline aya/src/maps/mod.rs:303]: MapError::InvalidPinPath should go

**Review by alessandrod — COMMENTED** [review]
I think `MapError::PinError` and `ProgramError::PinError` shouldn't exist, and we should instead return `PinError` directly from the pin methods.

In general we need to stop adding variants to existing errors, we should be removing them instead. Imagine you have:

```
match bpf.map_something_that_doesnt_do_any_pinning() {
   MapError::SomethingINeedToHandle => ...,
   MapError::SomethingElseINeedToHandle => ...,
   MapError::PinError(_) => I shouldn't have to handle this i'm not pinning,
   ...
}
```

**Pull request merged**